avro-codegen/src/main/java/com/linkedin/avro/codegen/CodeGenerator.java

```diff
@@ -0,0 +1,775 @@
+                        //not a redefinition, just a known schema
+                        continue;
+                    }
+                    throw new IllegalArgumentException("schema " + fullName + " is defined in " + schemaFile + " and also on the classpath");
+                }
+            }
+
+            actuallyNewSchemas.put(schema.getFullName(), schema);
+        }
+
+        return actuallyNewSchemas;
+    }
+
+    private void buildUpKnownSchemaLists(
+            Map<String, SchemaDetails> successfullyParsed,
+            Map<String, SchemaDetails> moreSchemas,
+            Map<String, SchemaDetails> externalSchemas,
+            List<Schema> allKnownSchemas,
+            List<Schema> allKnownTopLevelSchemas
+    ) {
+        successfullyParsed.forEach((fullName, schemaDetails) -> {
+            if (schemaDetails.topLevel) {
+                allKnownTopLevelSchemas.add(schemaDetails.schema);
+            }
+            allKnownSchemas.add(schemaDetails.schema);
+        });
+        if (moreSchemas != null) {
+            moreSchemas.forEach((fullName, schemaDetails) -> {
+                if (schemaDetails.topLevel) {
+                    allKnownTopLevelSchemas.add(schemaDetails.schema);
+                }
+                allKnownSchemas.add(schemaDetails.schema);
+            });
+        }
+        if (externalSchemas != null) {
+            externalSchemas.forEach((fullName, schemaDetails) -> allKnownSchemas.add(schemaDetails.schema));
+        }
+    }
+
+    private void findAllIncludedFiles(Map<File, File> fileToParent, Set<File> sharedSchemas, Set<File> nonSharedSchemas) {
+        String[] includesArray = includes.toArray(new String[0]);
+
+        for (File input : inputs) {
+            if (!input.exists()) {
+                LOG.warn("path {} does not exist", input);
+            }
+
+            if (input.isFile()) {
+                //individual files are always considered included, so no test against includes
+                includeFile(input, input, fileToParent, sharedSchemas, nonSharedSchemas);
+            } else if (input.isDirectory()) {
+                DirectoryScanner scanner = new DirectoryScanner();
+                scanner.setBasedir(input);
+                scanner.setIncludes(includesArray);
+                scanner.scan();
+                //results are paths relative (==under) input
+                String[] includedFiles = scanner.getIncludedFiles();
+                for (String included : includedFiles) {
+                    File file = new File(input, included);
+                    includeFile(file, input, fileToParent, sharedSchemas, nonSharedSchemas);
+                }
+            } else {
+                LOG.warn("path {} is neither a file nor a directory", input);
+            }
+        }
+    }
+
+    private void includeFile(File file, File root, Map<File, File> fileToParent, Set<File> sharedSchemas, Set<File> nonSharedSchemas) {
```

Comment:
If I'm understanding this correctly, I think `fileToParent` might be more clearly named as `fileToRoot`?